charger/homeassistant-switch.go

```diff
@@ -0,0 +1,130 @@
+package charger
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/evcc-io/evcc/api"
+	"github.com/evcc-io/evcc/util"
+)
+
+type HomeAssistantSwitch struct {
+	baseURL      string
+	token        string
+	switchEntity string
+	powerEntity  string
+	client       *http.Client
+	*switchSocket
+}
+
+func init() {
+	registry.Add("homeassistant-switch", NewHomeAssistantSwitchFromConfig)
+}
+
+type haSwitchConfig struct {
+	embed        `mapstructure:",squash"`
+	BaseURL      string
+	Token        string
+	SwitchEntity string
+	PowerEntity  string
+	StandbyPower float64
+}
+
+func NewHomeAssistantSwitchFromConfig(other map[string]interface{}) (api.Charger, error) {
+	var cc haSwitchConfig
+	if err := util.DecodeOther(other, &cc); err != nil {
+		return nil, err
+	}
+	return NewHomeAssistantSwitch(cc.embed, cc.BaseURL, cc.Token, cc.SwitchEntity, cc.PowerEntity, cc.StandbyPower)
+}
+
+func NewHomeAssistantSwitch(embed embed, baseURL, token, switchEntity, powerEntity string, standbypower float64) (*HomeAssistantSwitch, error) {
+	c := &HomeAssistantSwitch{
+		baseURL:      baseURL,
+		token:        token,
+		switchEntity: switchEntity,
+		powerEntity:  powerEntity,
+		client:       &http.Client{Timeout: 5 * time.Second},
```

Comment:
Looks like your using "plain" Go. In evcc we have a bit of additional comfort. I'd suggest to use

    request.NewHelper 

here plus an `http.Transport` with headers decorator. This will simplify code and error handling.